toolchain/build-toolchain.sh

```diff
@@ -30,76 +45,151 @@ GCC=gcc-9.3.0
 TARGET="$1"
 BUILD_ROOT="$2"
 TOOLCHAIN_PREFIX="${BUILD_ROOT}/${TARGET}"
-MAKEOPTS="-j9 -s"
+MAKEFLAGS="-j9 -s"
+export MAKEFLAGS
 
-log "downloading packages"
 mkdir -p "${TOOLCHAIN_PREFIX}"
 cp ./*.patch "${BUILD_ROOT}"
 cd "${BUILD_ROOT}"
 
-[ ! -f ${BINUTILS}.tar.bz2 ] && wget "http://ftp.gnu.org/gnu/binutils/${BINUTILS}.tar.bz2"
-[ ! -f ${GCC}.tar.xz ] && wget "http://www.mirrorservice.org/sites/ftp.gnu.org/gnu/gcc/${GCC}/${GCC}.tar.xz"
+download() {
+    log "downloading packages"
 
-log "extracting packages"
-[ ! -d ${BINUTILS} ] && tar jxf ${BINUTILS}.tar.bz2
-[ ! -d ${GCC} ] && tar Jxf ${GCC}.tar.xz
+    [ ! -f ${BINUTILS}.tar.bz2 ] && wget "http://ftp.gnu.org/gnu/binutils/${BINUTILS}.tar.bz2"
+    [ ! -f ${GCC}.tar.xz ] && wget "http://www.mirrorservice.org/sites/ftp.gnu.org/gnu/gcc/${GCC}/${GCC}.tar.xz"
 
+    log "extracting packages"
+    [ ! -d ${BINUTILS} ] && tar jxf ${BINUTILS}.tar.bz2
+    [ ! -d ${GCC} ] && tar Jxf ${GCC}.tar.xz
 
-log "patching ${BINUTILS}"
-for patchfile in "${BINUTILS}"-*.patch; do
-    if [ ! -f "${BINUTILS}/$patchfile.applied" ]; then
-        patch -d "${BINUTILS}" -p1 < "$patchfile"
-        touch "${BINUTILS}/$patchfile.applied"
-    fi
-done
+    log "downloading GCC dependencies"
+    (cd "$GCC" && ./contrib/download_prerequisites)
+}
 
-log "building binutils"
-rm -rf "${BINUTILS}/build"
-mkdir -p "${BINUTILS}/build"
-cd "${BINUTILS}/build"
+build_binutils() {
+    log "patching ${BINUTILS}"
+    for patchfile in "${BINUTILS}"-*.patch; do
+        if [ ! -f "${BINUTILS}/$patchfile.applied" ]; then
+            patch -d "${BINUTILS}" -p1 < "$patchfile"
+            touch "${BINUTILS}/$patchfile.applied"
+        fi
+    done
+
+    log "building binutils"
+    rm -rf "${BINUTILS}/build"
+    mkdir -p "${BINUTILS}/build"
+    pushd "${BINUTILS}/build" > /dev/null
+
+    ../configure --target="${TARGET}" --prefix="${TOOLCHAIN_PREFIX}" \
+                 --with-sysroot="${TOOLCHAIN_PREFIX}/${TARGET}" --enable-lto
+    make
+
+    log "installing binutils"
+    make install
+    popd > /dev/null
+}
 
-../configure --target="${TARGET}" --prefix="${TOOLCHAIN_PREFIX}" \
-             --with-sysroot="${TOOLCHAIN_PREFIX}/${TARGET}" --enable-lto
-make ${MAKEOPTS}
+build_gcc_stage1() {
+    log "patching ${GCC}"
+    for patchfile in "${GCC}"-*.patch; do
+        if [ ! -f "${GCC}/$patchfile.applied" ]; then
+            patch -d "${GCC}" -p1 < "$patchfile"
+            touch "${GCC}/$patchfile.applied"
+        fi
+    done
+
+    log "building GCC (stage1)"
+    rm -rf "${GCC}/build"
+    mkdir -p "${GCC}/build"
+    cd "${GCC}/build"
```

Comment:
Maybe `pushd` and `popd` as in `buid_binutils()`

Reply:
changed to pushd/popd for symmetry